### PR TITLE
Feature/plugin

### PR DESCRIPTION
**Pull request by d34dman**
Merge request raised against #53 since the work is on top of it.

1. Introduced support for plugin architecture using yapsy - http://yapsy.sourceforge.net/
2. Added an example `Foo` plugin
3. This is a proof of concept.
4. `fab foo`, `fab foobar` should both be available, thanks for aliases.
5. In case aliases is not provided by implementing class, then `ClassName` would be the name of the task.

### CODE REVIEW

**Comment by stmh** [comment]
what are you trying to accomplish? Please provide a use-case.

**Comment by stmh** [comment]
I like the idea of extending fabalicious easily with custom code. I tried to modularize fabalicious as good as possible. it would be great if you can extend fabalicious on a per project level or provide a custom method (similar to the platform-method) only for a given project, without changing fabalicious core.

But this will require more work on top of your merge-request, and on the other hand fabric 2.0 got released, which is not compatible with fabalicious. Not to mention the python3 vs python2 hell.

**Comment by d34dman** [comment]
> what are you trying to accomplish? Please provide a use-case.

- Possibility to extend fabalicious without hacking core. You could write a new task and expose it in fabalicious without hacking `fabfily.py`.
- Plugin discovery : you could set up plugin discovery places, right now it is configured to look into just one place. But you could pass more directories to search into. @see `manager.setPluginPlaces([root_folder + '/plugins'])`.

> I like the idea of extending fabalicious easily with custom code. I tried to modularize fabalicious as good as possible. it would be great if you can extend fabalicious on a per project level or provide a custom method (similar to the platform-method) only for a given project, without changing fabalicious core.

As discussed in Plugin discovery section above, we could decide on the folder names. One of them could be `.tools/fabalicious` too. I have not tested what happens when same alias is provided by plugin which is also provided by core. Need to check if it is possible to override the core task.

> But this will require more work on top of your merge-request, 

If i understand correctly, we just need to decide the folders to search for custom plugins.

> and on the other hand fabric 2.0 got released, which is not compatible with fabalicious. Not to mention the python3 vs python2 hell.

After reading the changes and other things, i am little afraid about migrating right away. However there is a yapsy package for python3.

**Comment by d34dman** [comment]
I have updated PR to search for plugins in "~/.fabalicious/plugins" and as well as directory specified in fabfile.yaml.
To specify directory via fabfile.yaml, add `plugins_dir: ".tools/fabalicious/plugins"` to it.
Then keep your plugins inside `.tools/fabalicious/plugins` folder relative to project root.

**Comment by d34dman** [comment]
@stmh  right now the PR assumes all plugins to be of type Task. I would change this behaviour to look for a specific plugin of category Task, this way we can use the same plugin system to load non Task plugins in the future.

**Comment by d34dman** [comment]
Ok, now we look for a specifically, instances of `ITaskPlugin` to dynamically load tasks.

**Comment by stmh** [comment]
Hi @d34dman 

I extended the new plugin-mechanism so it can handle custom methods similar to custom tasks. I needed to refactor the global settings handling, I moved the settings into their methods, instead of handling them globally.

File: lib/plugins/__init__.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      result[plugin.name] = plugin.plugin_object
+  return result
+
+
+def getTasks(root_folder):
+  try:
+    __import__('imp').find_module('yapsy')
+    from task import ITaskPlugin
+
+    return init(root_folder, 'customTasksFolder', 'tasks', 'task', { "Task": ITaskPlugin })
```
Comment (d34dman) [inline lib/plugins/__init__.py]: NOTE: Just a discussion, not asking you to implement it.

yapsy searches into sub-folder too, so do we need to specify Task specific folder? 

The advantage of current solution is some time saved during plugin discovery.

The tasks will be filtered based on the class it extends. So this gives a good degree of freedom on the folder structure that developer could follow to organise his plugins.

Possible solution
1. use `customPluginsFolder` and add subfolders `task` and `methods` inside to keep it organised.
2. use `customPluginsFolders` (notice the `s`), and accept an array of locations that should be searched for.
3. In normal use case, the developer writing plugins would have to add the entry in fabfile.yaml just once.

Thinking about it further, lets  alter the key a little more and make it,

```
customPlugins:
  folder: ".fabalicious/plugins"
```
We might want to introduce/capture few more parameters in the future.

File: lib/plugins/__init__.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from os.path import expanduser
+from lib import configuration
+import logging
+log = logging.getLogger('fabric.fabalicious.plugins')
+
+def init(root_folder, settings_name, folder_name, plugin_name, categories_filter):
```
Comment (d34dman) [inline lib/plugins/__init__.py]: `init` is returning an array of plugins, How about we call it `loadPlugins` or `getPlugins`

Comment (d34dman) [reply lib/plugins/__init__.py]: `root_folder` is being passed around so much, can we put it into `configuration` ?

Somewhere in `fabfile.py`, it is being set as `configuration.fabfile_basedir = root_folder`, however, when you try to read `configuration.fabfile_basedir`, it gives back the path where `fabfile.yaml` is present. Counter intuitive.

**Pull request merged**